File: server/handlers/translationCard.go
package handlers

import (
	"encoding/json"
	"languageNotes/service"
	"net/http"
	"strconv"

	"github.com/jackc/pgx/v5/pgxpool"
	"go.uber.org/zap"
)

type TranslationHandler struct {
	conn *pgxpool.Pool
}

type postTranslationRequestBody struct {
	English  string `json:"english"`
	Language string `json:"language"`
	SetID    int    `json:"setId"`
}

type patchTranslationCardRequestBody struct {
	English    string `json:"english"`
	Meaning    string `json:"meaning"`
	Translated string `json:"translated"`
}

func NewTranslationHandler(p *pgxpool.Pool) *TranslationHandler {
	return &TranslationHandler{conn: p}
}

// POST /translate
//
//	request: {
//		english: string
//		language: string
//	}
//
// response: No body only 200
// This API will translate an English version of a word to the desired language, then it will add the
// respective data to the database
func (h *TranslationHandler) PostTranslate(w http.ResponseWriter, r *http.Request) {
	// Create a logger
	logger, err := createLoggerForRequest(r)

	if err != nil {
		http.Error(w, "Internal server error", http.StatusInternalServerError)
		return
	}

	// Retrieve UUID from context
	uuid, err := retrieveUUIDfromContext(r)
	if err != nil {
		http.Error(w, err.Error(), http.StatusBadRequest)
		return
	}

	// Entry log
	logger.Info(
		"PostTranslate",
		zap.String("uuid", uuid),
	)

	var reqBody *postTranslationRequestBody

	if err := json.NewDecoder(r.Body).Decode(&reqBody); err != nil {
		http.Error(w, "Invalid request", http.StatusBadRequest)
		return
	}

	// TODO: Validate request

	s := service.NewTranslationService(h.conn, r.Context(), logger)

	err = s.CreateNewTranslationCard(uuid, reqBody.English, reqBody.Language, reqBody.SetID)

	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}

	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(http.StatusOK)
	// TODO: Maybe add a return body so that we know what was added?
}

// GET /set/{setID}/translation-cards
func (h *TranslationHandler) GetTranslationCards(w http.ResponseWriter, r *http.Request) {
	// Create a logger
	logger, err := createLoggerForRequest(r)

	if err != nil {
		http.Error(w, "Internal server error", http.StatusInternalServerError)
		return
	}

	// Retrieve UUID from context
	uuid, err := retrieveUUIDfromContext(r)
	if err != nil {
		http.Error(w, err.Error(), http.StatusBadRequest)
		return
	}

	// Entry log
	logger.Info(
		"GetTranslationCards",
		zap.String("uuid", uuid),
	)

	// Retrieve values from the path
	setID, err := strconv.Atoi(r.PathValue("setID"))

	if err != nil {
		http.Error(w, err.Error(), http.StatusBadRequest)
		return
	}

	// TODO: Validate request

	// Create new service and call the relevant method
	s := service.NewTranslationService(h.conn, r.Context(), logger)

	res, err := s.TranslationCards(uuid, setID)

	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}

	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(http.StatusOK)
	json.NewEncoder(w).Encode(res)
}

func (h *TranslationHandler) PatchTranslationCard(w http.ResponseWriter, r *http.Request) {
	// Create a logger
	logger, err := createLoggerForRequest(r)

	if err != nil {
		http.Error(w, "Internal server error", http.StatusInternalServerError)
		return
	}

	// Retrieve UUID from context
	uuid, err := retrieveUUIDfromContext(r)
	if err != nil {
		http.Error(w, err.Error(), http.StatusBadRequest)
		return
	}

	// Entry log
	logger.Info(
		"PatchTranslationCard",
		zap.String("uuid", uuid),
	)

	// Retrieve values from the path
	cardID, err := strconv.Atoi(r.PathValue("cardID"))

	if err != nil {
		http.Error(w, err.Error(), http.StatusBadRequest)
		return
	}

	// TODO: Validate request

	var reqBody *patchTranslationCardRequestBody

	if err := json.NewDecoder(r.Body).Decode(&reqBody); err != nil {
		http.Error(w, "Invalid request", http.StatusBadRequest)
		return
	}

	// Create new service and call the relevant method
	s := service.NewTranslationService(h.conn, r.Context(), logger)

	err = s.UpdateTranslationCard(uuid, cardID, reqBody.English, reqBody.Meaning, reqBody.Translated)

	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}

	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(http.StatusOK)
}

func (h *TranslationHandler) DeleteTranslationCard(w http.ResponseWriter, r *http.Request) {
	// Create a logger
	logger, err := createLoggerForRequest(r)

	if err != nil {
		http.Error(w, "Internal server error", http.StatusInternalServerError)
		return
	}

	// Retrieve UUID from context
	uuid, err := retrieveUUIDfromContext(r)
	if err != nil {
		http.Error(w, err.Error(), http.StatusBadRequest)
		return
	}

	// Entry log
	logger.Info(
		"DeleteTranslationCard",
		zap.String("uuid", uuid),
	)

	// Retrieve value from the path
	cardID, err := strconv.Atoi(r.PathValue("cardID"))

	s := service.NewTranslationService(h.conn, r.Context(), logger)

	if err := s.DeleteTranslationCard(cardID); err != nil {
		http.Error(w, err.Error(), http.StatusBadRequest)
		return
	}
}

File: frontend/language-notes/src/api/translationCard.ts
export async function fetchTranslationCards(setId: number, token: string): Promise<TranslationCard[]> {
    const url = `${import.meta.env.VITE_SERVER_URL}/set/${setId}/translation-cards`;
  
    try {
      const response = await fetch(url, {
        method: "GET", 
        headers: {
          "Content-Type": "application/json",
          "Authorization": `Bearer ${token}`
        }
      });
  
      if (!response.ok) {
        throw new Error(`HTTP error! Status: ${response.status}`);
      }
  
      const translationCards = await response.json();
  
      return translationCards;
    } catch (error) {
      console.error("Error fetching translation cards:", error);
      throw error;
    }
};

export async function translateEnglishWord(english: string, language: string, token: string, setId: number) {
  await fetch("${import.meta.env.VITE_SERVER_URL}/translate", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      "Authorization": `Bearer ${token}`
    },
    body: JSON.stringify({
      english: english,
      language: language,
      setId: setId
    }),
  });
  // TODO: Response handling currently no response from backend
  // const data = await response.json();
  // console.log(data)
  // return {
  //   translated: data.translated,
  //   meaning: data.meaning,
  // };
};

export async function updateTranslationCard(english: string, meaning: string, translated: string, cardId: number, token: string) {
  await fetch(`${import.meta.env.VITE_SERVER_URL}/translation-card/${cardId}`, {
    method: "PATCH",
    headers: {
      "Content-Type": "application/json",
      "Authorization": `Bearer ${token}`
    },
    body: JSON.stringify({
      english: english,
      meaning: meaning,
      translated: translated
    }),
  })
}

export async function deleteTranslationCard(cardId: number, token: string) {
  await fetch(`${import.meta.env.VITE_SERVER_URL}/translation-card/${cardId}`, {
    method: "DELETE",
    headers: {
      "Content-Type": "application/json",
      "Authorization": `Bearer ${token}`
    }
  });
}